File: android/src/main/java/com/cloudwebrtc/webrtc/detection/MotionDetection.kt
package com.cloudwebrtc.webrtc.detection

import android.os.Handler
import android.os.Looper
import android.util.Log
import com.cloudwebrtc.webrtc.video.LocalVideoTrack
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel
import org.webrtc.VideoFrame
import kotlin.concurrent.thread

class MotionDetection(binaryMessenger: BinaryMessenger) :
    LocalVideoTrack.ExternalVideoFrameProcessing,
    EventChannel.StreamHandler {
    private var videoTrack: LocalVideoTrack? = null
    private val pixelDetection by lazy { PixelDetection() }
    private val eventChannel =
        EventChannel(binaryMessenger, "FlutterWebRTC/motionDetection")
    private var eventSink: EventChannel.EventSink? = null
    private var prevDetection = 0L
    private var detectionLevel = 2
    private var intervalMs = 300
    private var listener: Listener? = null

    init {
        eventChannel.setStreamHandler(this)
    }

    fun setVideoTrack(videoTrack: LocalVideoTrack) {
        if (this.videoTrack != null) {
            return
        }
        this.videoTrack = videoTrack
        if (isActive) {
            startDetection(videoTrack)
        }
    }

    fun removeVideoTrack(trackId: String) {
        if (videoTrack?.id() != trackId) {
            return
        }
        stopDetection()
        videoTrack = null
    }

    private var isActive = false

    fun requestMotionDetection(request: DetectionRequest) {
        detectionLevel = request.level
        if (request.enabled == isActive) {
            return
        }
        isActive = request.enabled
        if (isActive) {
            videoTrack?.let { track ->
                startDetection(track)
            } ?: run {
                Log.d("Motion detection", "VideoTrack is null")
            }
        } else {
            stopDetection()
        }
    }

    fun addListener(listener: Listener) {
        this.listener = listener;
    }

    fun removeListener() {
        this.listener = null
    }

    private fun startDetection(videoTrack: LocalVideoTrack) {
        videoTrack.addProcessor(this)
        Log.d("Motion detection", "Motion detection started")
    }

    private fun stopDetection() {
        videoTrack?.removeProcessor(this)
        pixelDetection.resetPrevious()
        Log.d("TAG", "Motion detection stopped")
    }

    val frameIntervalMs: Long
        get() = this.intervalMs.toLong()

    override fun onFrame(videoFrame: VideoFrame): VideoFrame {
        if (System.currentTimeMillis() - prevDetection < intervalMs) {
            return videoFrame
        }
        prevDetection = System.currentTimeMillis()
        videoFrame.retain()
        val i420Buffer = videoFrame.buffer.toI420() ?: return videoFrame
        val rotation = videoFrame.rotation
        videoFrame.release()
        thread {
            pixelDetection.detect(
                buffer = i420Buffer,
                rotation = rotation,
                detectionLevel = detectionLevel
            ) { result ->
                sendDetection(result)
                listener?.onDetect(result)
            }
        }
        return videoFrame
    }

    private fun sendDetection(detected: DetectionFrame) {
        if (!isActive) return
        val params = detected.toMap()
        Handler(Looper.getMainLooper()).post {
            eventSink?.success(params)
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        this.eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        this.eventSink = null
    }

    fun dispose() {
        if (isActive) {
            stopDetection()
        }
        videoTrack = null
        eventSink = null
        eventChannel.setStreamHandler(null)

    }

    interface Listener {
        fun onDetect(detection: DetectionFrame)
    }

}

File: ios/Classes/MotionDetection.swift
//
//  MotionDetection.swift
//  flutter_webrtc
//
//  Created by MacBook 16 on 28.05.2022.
//

import Foundation
import Flutter
import CoreMedia
import CoreVideo
import AVFoundation

public class MotionDetection: NSObject {

    private lazy var pixelDetection = { PixelDetection() }()
    private var detectionLevel = 0
    private let eventChannel: FlutterEventChannel
    private var eventSink: FlutterEventSink?
    private var listener: MotionDetectionListener?
    private let log = Log(subsystem: "MotionDetection", category: "")
    private var active = false

    @objc public init(binaryMessenger: FlutterBinaryMessenger) {
        eventChannel = FlutterEventChannel(
            name: "FlutterWebRTC/motionDetection",
            binaryMessenger: binaryMessenger)
        super.init()
        eventChannel.setStreamHandler(self)
    }

    @objc public func setDetection(request: DetectionRequest) {
        log.d("setDetection \(request.enabled)")
        self.detectionLevel = request.level
        guard self.active != request.enabled else { return }
        self.active = request.enabled
    }

    @objc public func stop() {
        self.pixelDetection.resetPrevious()
        log.d("stopped")
    }

    @objc public func dispose() {
        if active {
            stop()
        }
        eventChannel.setStreamHandler(nil)
        eventSink = nil
        log.d("disposed")
    }

    func addListener(_ listener: MotionDetectionListener) {
        self.listener = listener
    }

    func removeListener() {
        self.listener = nil
    }

    func processFrame(buffer: UnsafePointer<UInt8>,
                      rotation: Int,
                      width: Int,
                      height: Int,
                      strideY: Int) {
        DispatchQueue.global(qos: .background).async { [weak self] in
            guard let self = self else { return }
            self.pixelDetection.detect(
                buffer: buffer,
                rotation: rotation,
                detectionLevel: self.detectionLevel,
                width: width,
                height: height,
                strideY: strideY,
                result: { [weak self] detected in
                    self?.sendDetectionResult(detected)
                    if !detected.detectedList.isEmpty {
                        self?.listener?.onDetected(result: detected)
                    }
                }
            )
        }
    }

    private func sendDetectionResult(_ result: DetectionFrame) {
        guard self.active else { return }
        let param: [String: Any] = result.toMap()
        DispatchQueue.main.async { [weak self] in
            guard let eventSink = self?.eventSink else {
                return
            }
            eventSink(param)
        }
    }
}

extension MotionDetection: FlutterStreamHandler {
    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        eventSink = events
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        eventSink = nil
        return nil
    }
}

protocol MotionDetectionListener {
    func onDetected(result: DetectionFrame)
}
